src/rust/ide/ast/impl/src/crumbs.rs

```diff
@@ -98,43 +142,167 @@ pub enum InfixCrumb {
+    HeadLine,
+    /// Index in the sequence of "rest of" lines (not counting the HeadLine).
+    TailLine {tail_index:usize},
 }
-impl From<&ModuleCrumb> for Crumb {
-    fn from(crumb: &ModuleCrumb) -> Self {
-        Crumb::Module(*crumb)
-    }
+
+
+// === Import ===
+
+#[allow(missing_docs)]
+#[derive(Clone,Copy,Debug,PartialEq,Hash)]
+pub struct ImportCrumb {pub index:usize}
+
+
+// === Mixfix ===
+
+#[allow(missing_docs)]
+#[derive(Clone,Copy,Debug,PartialEq,Hash)]
+pub enum MixfixCrumb {
+    Name {index:usize},
+    Args {index:usize}
+}
+
+
+// === Group ===
+
+#[allow(missing_docs)]
+#[derive(Clone,Copy,Debug,PartialEq,Hash)]
+pub struct GroupCrumb;
+
+
+// === Def ===
+
+#[allow(missing_docs)]
+#[derive(Clone,Copy,Debug,PartialEq,Hash)]
+pub enum DefCrumb {
+    Name,
+    Args {index:usize},
+    Body
 }
 
-impl From<InfixCrumb> for Crumb {
-    fn from(crumb: InfixCrumb) -> Self {
-        Crumb::Infix(crumb)
+
+// === Conversion Traits ===
+
+macro_rules! from_crumb {
+    ($id:ident) => {
+        paste::item! {
+            impl From<[<$id Crumb>]> for Crumb {
+                fn from(crumb:[<$id Crumb>]) -> Self {
+                    Crumb::$id(crumb)
+                }
+            }
+        }
+
+        paste::item! {
+            impl From<&[<$id Crumb>]> for Crumb {
+                fn from(crumb:&[<$id Crumb>]) -> Self {
+                    Crumb::$id(crumb.clone())
+                }
+            }
+        }
     }
 }
 
-impl From<&InfixCrumb> for Crumb {
-    fn from(crumb: &InfixCrumb) -> Self {
-        Crumb::Infix(*crumb)
+macro_rules! impl_crumbs {
+    ($($id:ident),*) => {
+        $(from_crumb!{$id})*
+
+        impl Crumbable for Shape<Ast> {
+            type Crumb = Crumb;
+            fn get(&self, crumb:&Self::Crumb) -> FallibleResult<&Ast> {
+                match (self,crumb) {
+                    $((Shape::$id(shape),Crumb::$id(crumb)) => shape.get(crumb),)*
+                    _ => Err(MismatchedCrumbType.into())
+                }
+            }
+
+            fn set(&self, crumb:&Self::Crumb, new_ast:Ast) -> FallibleResult<Self> {
+                match (self,crumb) {
+                    $((Shape::$id(shape),Crumb::$id(crumb)) => Ok(shape.set(crumb,new_ast)?.into()),)*
+                    _ => Err(MismatchedCrumbType.into())
+                }
+            }
+
+            fn iter_subcrumbs<'a>(&'a self) -> Box<dyn Iterator<Item = Self::Crumb> + 'a> {
+                match self {
+                    $(Shape::$id(shape) => Box::new(shape.iter_subcrumbs().map(Crumb::$id)),)*
+                    _ => Box::new(std::iter::empty())
+                }
+            }
+        }
+
+        paste::item!{
+            /// Crumb identifies location of child AST in an AST node. Allows for a single step AST traversal.
+            #[derive(Clone,Copy,Debug,PartialEq,Hash)]
+            #[allow(missing_docs)]
+            pub enum Crumb {
+                $($id([<$id Crumb>]),)*
+            }
+        }
```

Comment:
Please avoid using `paste`. 
Especially strongly avoid using it when introducing new datatypes, like `Crumb` here. It breaks code completion and navigation and is generally PITA to deal with later.

It's better to pass both `Shape` and `CrumbShape` arguments into the macro invocation, rather than trying to construct the latter from the `Shape`.

Here my suggestion is:
* introduce a macro that stores list of Shape and CrumbShape pairs;
* `from_crumb` and `impl_crumbs` should take that list — and then they don't need to use paste anymore